openeogeotrellis/logs.py

```diff
@@ -1,67 +1,136 @@
-                    {'log.offset': {'order': 'asc'}},  # tie-breaker
-                    # chances are slim that two processes write on the same line within the same millisecond
-                    # 'log.file.path', 'host.name'
-                ],
-                request_timeout=60
-            )
+    :param job_id:
+        ID of the Job
+
+    :param create_time:
+        Time the job was created, only log records starting from that time will be retrieved
 
-            hits = search_result['hits']['hits']
+    :param search_after:
+        Search only after this offset.
+        This offset is a list which takes two elements: the timestamp and log.offset
+        where timestamp is the Unix Epoch (int) and log.offset (int) in Elasticsearch.
 
-            for hit in hits:
-                search_after = hit['sort']
+        For example: [1673351608383, 102790]
 
-                # Skip the log line if the log level is empty
-                entry = _as_log_entry(log_id=json.dumps(search_after), hit=hit)
-                if "level" in entry:
-                    yield entry
+    :raises OpenEOApiException:
+        - Either when the offset is not valid JSON
+        - or when Elasticsearch had a connection timeout
 
-            if len(hits) < page_size:
-                break
+    :return: an generator that yields a dict for each log record.
+    """
+
+    page_size = 100
+    query = {
+        "bool": {
+            "filter": [
+                {"term": {"job_id": job_id}},
+                {"terms": {"tags": ES_TAGS}},
+            ],
+        }
+    }
+    if create_time:
+        query["bool"]["filter"].append(
+            {
+                "range": {
+                    "@timestamp": {
+                        "format": "strict_date_optional_time",
+                        "gte": rfc3339.datetime(create_time),
+                    }
+                }
+            }
+        )
+    with Elasticsearch(ES_HOSTS) as es:
+        while True:
+            try:
+                search_result = es.search(
+                    index=ES_INDEX_PATTERN,
+                    query=query,
+                    search_after=search_after,
+                    size=page_size,
+                    sort=[
+                        {"@timestamp": {"order": "asc"}},
+                        {"log.offset": {"order": "asc"}},  # tie-breaker
+                        # chances are slim that two processes write on the same line within the same millisecond
+                        # 'log.file.path', 'host.name'
+                    ],
+                    request_timeout=120,
+                )
+
+            except ConnectionTimeout as exc:
+                message = (
+                    "Temporary failure while retrieving logs (ConnectionTimeout). "
+                    + "Please try again and report this error if it persists."
+                )
+                raise OpenEOApiException(status_code=504, message=message)
```

Comment:
We're losing valuable debugging information here i.e. the underlying ReadTimeoutError; is it possible to somehow retain this?

Reply:
Indeed it would be better to keep the info of the original exception somehow.

I will see if I can get and log that req_id from the view there, without too many complications.
I was also thinking we could log the entire ConnectionTimeout  as a warning. Maybe that is enough.
Generally, we have to be a bit careful with logging stuff inside code that processes logging, but to me it looks like that would be OK .

Additionally, would it be useful to use a `raise from` to link the original exception to the OpenEOApiException?
I suspect we should in fact avoid that, because I think it could still leak the security sensitive info that way.
While it may not be directly visible in the OpenEOApiException's message, the info is still present, and could be accessed.


Perhaps, something like this?

```python
            except ConnectionTimeout as exc:
                _logger.warning("Connection timeout while retrieving logs from Elasticsearch", exc_info=True)
                message = (
                    "Temporary failure while retrieving logs (ConnectionTimeout). "
                    + "Please try again and report this error if it persists."
                )
                raise OpenEOApiException(status_code=504, message=message) from exc
```

Reply:
The idea was that we log the details but without a `job_id`; otherwise a user would simply be able to request the logs for this job ID again and still see the sensitive information.

I would include the request ID (`req_id`) because then we are able to take it from the user's error message and look up said log in Kibana.

> Generally, we have to be a bit careful with logging stuff inside code that processes logging, but to me it looks like that would be fine here.

In this case the logs that we are requesting (with a `job_id`) and the one that we write (without a `job_d`) will be mutually exclusive so I don't expect any issues.

Now that you mention it: maybe an easy fix without passing on the `req_id` all the way is to just chain the exceptions. If I understand this piece of code correctly, the chain of exceptions will already be logged with a `req_id` and the error message to the user will not include the `ConnectionTimeout` with the sensitive info:

https://github.com/Open-EO/openeo-python-driver/blob/9a1792ac98e33a37ecea681aedc00c2e3fef8eca/openeo_driver/views.py#L1236-L1240

Reply:
Now I'm chaining the exception. 
Also added the req_id, it makes it easier to recognize when it is also in the message for the user.

There is a new test in `test_views.py, test_get_job_logs_with_connection_timeout`, which verifies a few important points about the logging, namely:
- the host URL is not leaked in the message for the user,
- the message format is the user friendly one we expect,
- the general log (the one not about the job) still contains the traceback information of ConnectionTimeouts.

https://github.com/Open-EO/openeo-geopyspark-driver/blob/65b1356011a8b58f70c510379f8682ea8fca324a/tests/test_views.py#L932-L1006

Let me know what you think.

Reply:
Pushed some small changes myself, otherwise looked good.

- slightly changed error message to align it with other errors towards the client ("... (ref: r-b66a7c4f422440f6bf50db8b1d4fac0f)")
- more robust tests, in particular the last assert ("ConnectionTimeout" would also be present in the user's error message so a false negative)